hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -622,7 +623,13 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: I
     val rightKeyFieldIdx = rightRVDType.kFieldIdx
     val leftValueFieldIdx = leftRVDType.valueFieldIdx
     val rightValueFieldIdx = rightRVDType.valueFieldIdx
-    val newRowPType = PType.canonical(newRowType).asInstanceOf[PStruct]
+
+    val leftKeyType = leftRVDType.kType
+    val leftValueType = leftRVDType.valueType
+    val rightValueType = rightRVDType.valueType
+
+    val newRowPType = leftKeyType ++ leftValueType ++ rightValueType
```

Comment:
this is correct for inner join, but not left, right, or outer.

Reply:
one sec, changing to match to ensure always exhaustive